xdsl/dialects/stencil.py

```diff
@@ -731,6 +755,66 @@ def verify_(self) -> None:
             )
 
 
+@dataclass(frozen=True)
+class AccessPattern:
+    """
+    Represents access patterns of a stencil apply operation.
+
+    Contains a bunch of helpers to get common information about accesses.
+    """
+
+    accesses: tuple[tuple[int, ...], ...]
+
+    @property
+    def dims(self):
+        if not self.accesses:
+            return 0
+        return len(self.accesses[0])
+
+    @property
+    def is_diagonal(self) -> bool:
+        for _ in self.get_diagonals():
+            return True
+        return False
+
+    def get_diagonals(self) -> Iterable[tuple[int, ...]]:
+        for ax in self.accesses:
+            # number of nonzero offset > 1 => diagonal
+            if sum(1 if x != 0 else 0 for x in ax) > 1:
+                yield ax
+
+    def extent_in_axis(self, axis: int) -> tuple[int, int]:
+        left, right = 0, 0
+        for ax in self.accesses:
+            left = min(ax[axis], left)
+            right = max(ax[axis], right)
+        return left, right
+
+    def halos(self) -> tuple[tuple[int, int], ...]:
+        n = self.dims
+        lefts, rights = [0] * n, [0] * n
+        for ax in self.accesses:
+            for axis in range(n):
+                lefts[axis] = min(ax[axis], lefts[axis])
+                rights[axis] = max(ax[axis], rights[axis])
```

Comment:
This could reuse `extent_in_axis`

Reply:
I don't see it?
If you swap the loops, the inner one is exactly extent_in_axis, and the total operations are the same, or am I missing something?

Reply:
Oh, because one of those is element-wise max?

I would prefer readability before this complexity becomes a problem, I'm not sure we'll ever have asymptotically big numbers of accesses.

But it's nitpicking 🙂